temporal-sdk/src/main/java/io/temporal/internal/worker/Poller.java

```diff
@@ -232,27 +237,44 @@ public void run() {
           suspender.await();
         }
 
-        if (pollExecutor.isShutdown()) {
-          return;
-        }
+        interruptIfNeeded();
+
         task.run();
         pollBackoffThrottler.success();
+      } catch (InterruptedException e) {
+        // we restore the flag here, so it can be checked and processed (with exit) in finally
+        Thread.currentThread().interrupt();
       } catch (Throwable e) {
         pollBackoffThrottler.failure();
-        if (!pollExecutor.isShutdown() && !pollExecutor.isTerminating()
-            || !(e.getCause() instanceof InterruptedException)
-                && !(e instanceof RejectedExecutionException)) {
+        // if we are terminating and getting rejected execution - it's normal
+        if (!pollExecutor.isTerminating() || !(e instanceof RejectedExecutionException)) {
           uncaughtExceptionHandler.uncaughtException(Thread.currentThread(), e);
         }
       } finally {
-        // Resubmit itself back to pollExecutor
-        if (!pollExecutor.isTerminating() && !pollExecutor.isShutdown()) {
+        if (!shouldInterrupt()) {
+          // Resubmit itself back to pollExecutor
           pollExecutor.execute(this);
         } else {
-          log.info("poll loop done");
+          log.info("terminated; poll loop done");
         }
       }
     }
+
+    private void interruptIfNeeded() throws InterruptedException {
+      if (shouldInterrupt()) {
+        throw new InterruptedException();
+      }
+    }
+
+    private boolean shouldInterrupt() {
+      boolean threadIsInterrupted = Thread.interrupted();
+      if (threadIsInterrupted) {
+        // it's a good idea to preserve the flag, because it's not obvious that this method will
+        // flush the flag
+        Thread.currentThread().interrupt();
+      }
+      return pollExecutor.isTerminating() || threadIsInterrupted;
```

Comment:
`pollExecutor.isTerminating()` is covering both `pollExecutor.isShutdown()` and `pollExecutor.isTerminating()`